instance/models/redis_server.py

```diff
@@ -0,0 +1,120 @@
+# Copyright (C) 2015-2021 OpenCraft <xavier@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app models - Redis Server model and manager
+"""
+
+# Imports #####################################################################
+
+import logging
+from urllib.parse import urlparse
+
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+from django.db import models
+from django_extensions.db.models import TimeStampedModel
+
+from instance.models.shared_server import SharedServerManager
+from instance.models.utils import ValidateModelMixin
+
+
+# Logging #####################################################################
+
+logger = logging.getLogger(__name__)
+
+
+# Constants ###################################################################
+
+DEFAULT_INSTANCE_REDIS_PORT = 5671
+DEFAULT_INSTANCE_REDIS_DB = 0
+
+
+# Models ######################################################################
+
+class RedisServerManager(SharedServerManager):
+    """
+    Custom manager for the RedisServer model.
+    """
+
+    def _create_default(self):
+        """
+        Create the default Redis server configured in the Django settings, if any.
+        """
+        if settings.DEFAULT_INSTANCE_REDIS_URL:
+            conn_params = urlparse(settings.DEFAULT_INSTANCE_REDIS_URL)
+
+            defaults = {
+                'admin_username': conn_params.username,
+                'admin_password': conn_params.password,
+                'instance_host': conn_params.hostname,
+                'instance_port': conn_params.port,
+                'instance_db': conn_params.path.replace("/", ""),
+                'use_ssl_connections': conn_params.scheme == "rediss",
+                'accepts_new_clients': True,
+            }
+
+            if not all(defaults.values()):
+                raise ImproperlyConfigured(
+                    "Found DEFAULT_INSTANCE_REDIS_URL, but it must contain "
+                    "basic auth credentials, hostname, port and db."
+                )
+
+            logger.info("Creating RedisServer %s", defaults['instance_host'])
+
+            defaults['name'] = defaults['instance_host']
+            defaults['instance_db'] = int(defaults['instance_db'])
+
+            server, created = self.get_or_create(defaults=defaults)
```

Comment:
can this get_or_create based on the hostname or something, so it doesn't need that random pk function (which I'm guessing that's what it's for)?

Reply:
The `get_or_create` is based on the connection parameters only (a.k.a the parsed URI/defaults). Since changing any parameter of the connections parameters should be represented as a new redis server (even if the db param is changed).

Reply:
Ok sure.  As long as the behaviour is documented - otherwise the relationship between `DEFAULT_INSTANCE_REDIS_URL` and an actual redis configuration in the database is confusing.

Reply:
This has been documented in a previous commit.